treesitter: default start and end row when omitted

jvgrootveld opened the pull request:
Add support for default start and end row when omitted in the
query:iter_captures and query:iter_matches functions.

When the start and end row values are omitted, the values of the given
node is used. The end row value is incremented by 1 to include the node end
row in the match.

Updated tests and docs accordingly.

See also #13761

Commit:
treesitter: default start and end row when omitted

Add support for default start and end row when omitted in the
query:iter_captures and query:iter_matches functions.

When the start and end row values are omitted, the values of the given
node is used. The end row value is incremented by 1 to include the node end
row in the match.

Updated tests and docs accordingly.

jvgrootveld commented:
@vigoux I have a linting warning because I wanted the implementation to be simple and small by directly replace the parameters. But because of that, the `_` variable is created globally. What is the nicest way to solve this? I thought about 2 solutions.

**Current implementation (with linting warning)**
```lua
start, _, stop, _ = node:range()
stop = stop + 1 -- Make stop inclusive
```

**Idea 1: Create a local placeholder for the values**
```lua
local node_start, _, node_stop, _ = node:range()
start = node_start
stop = node_stop + 1 -- Make stop inclusive
```

**Idea 2: Create a local unused variable for `_`**
*(This feels dirty though)*
```lua
local _
start, _, stop, _ = node:range()
stop = stop + 1 -- make stop inclusive
```

vigoux commented:
Idea 1 (with the local and new names) is the one I would suggest to use.

Thanks for this contribution though!

steelsojka commented on runtime/lua/vim/treesitter/query.lua:
```diff
@@ -353,6 +353,12 @@ function Query:iter_captures(node, source, start, stop)
     source = vim.api.nvim_get_current_buf()
   end
+
+  if start == nil and stop == nil then
```
Since these two chunks are identical it would be nice to keep it DRY and move this to a function.

jvgrootveld replied:
Good idea, together with *idea 1* this would be more clean.

Commit:
treesitter: Fix linter warning and add helper function to remove duplicated logic

This function returns the start and stop value if set else the node's range is used
When the node's range is used, the stop is incremented by 1 to make the search inclusive
